File: app-android/src/main/java/com/ustadmobile/port/android/view/SimplePeopleListRecyclerAdapter.java
package com.ustadmobile.port.android.view;

import android.arch.paging.PagedListAdapter;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.toughra.ustadmobile.R;
import com.ustadmobile.core.controller.CommonHandlerPresenter;
import com.ustadmobile.core.controller.UstadBaseController;
import com.ustadmobile.lib.db.entities.Person;

/**
 * A Simple recycler adapter for a dead simple list of Students.
 */

public class SimplePeopleListRecyclerAdapter extends
        PagedListAdapter<Person,
                SimplePeopleListRecyclerAdapter.ClazzStudentViewHolder> {

    Context theContext;
    Fragment theFragment;
    CommonHandlerPresenter mPresenter;


    protected class ClazzStudentViewHolder extends RecyclerView.ViewHolder {
        protected ClazzStudentViewHolder(View itemView) {
            super(itemView);
        }
    }

    protected SimplePeopleListRecyclerAdapter(@NonNull DiffUtil.ItemCallback<Person> diffCallback,
                                               Context context, Fragment fragment) {
        super(diffCallback);
        theContext = context;
        theFragment = fragment;
    }

    protected SimplePeopleListRecyclerAdapter(@NonNull DiffUtil.ItemCallback<Person> diffCallback,
                                              Context context, Fragment fragment,
                                              CommonHandlerPresenter presenter) {
        super(diffCallback);
        theContext = context;
        theFragment = fragment;
        mPresenter = presenter;
    }

    protected SimplePeopleListRecyclerAdapter(@NonNull DiffUtil.ItemCallback<Person> diffCallback,
                                              Context context) {
        super(diffCallback);
        theContext = context;
    }

    protected SimplePeopleListRecyclerAdapter(@NonNull DiffUtil.ItemCallback<Person> diffCallback,
                                              Context context, CommonHandlerPresenter presenter) {
        super(diffCallback);
        theContext = context;
        mPresenter = presenter;
    }

    /**
     * This method inflates the card layout (to parent view given) and returns it.
     * @param parent View given.
     * @param viewType View Type not used here.
     * @return New ViewHolder for the ClazzStudent type
     */
    @NonNull
    @Override
    public SimplePeopleListRecyclerAdapter.ClazzStudentViewHolder
            onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        View clazzStudentListItem =
                LayoutInflater.from(theContext).inflate(
                        R.layout.item_peoplelist, parent, false);
        return new SimplePeopleListRecyclerAdapter.ClazzStudentViewHolder(clazzStudentListItem);
    }

    /**
     * This method sets the elements after it has been obtained for that item'th position.
     * @param holder    The holder
     * @param position  The position in the recycler view.
     */
    @Override
    public void onBindViewHolder(
            @NonNull SimplePeopleListRecyclerAdapter.ClazzStudentViewHolder holder, int position) {


        Person thisPerson = getItem(position);
        String studentName;
        if(thisPerson == null){
            studentName = "Student";
        }else{
            studentName = thisPerson.getFirstNames() + " " +
                    thisPerson.getLastName();
        }

        TextView studentEntry = (TextView)holder.itemView
                .findViewById(R.id.item_peoplelist_name);
        studentEntry.setText(studentName);


        studentEntry.setOnClickListener(v -> mPresenter.handleCommonPressed(thisPerson.getPersonUid()));

    }
}

File: app-android/src/main/kotlin/com/ustadmobile/port/android/view/SimplePeopleListRecyclerAdapter.kt
package com.ustadmobile.port.android.view
import android.content.Context
import android.content.res.Resources
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso

import com.toughra.ustadmobile.R
import com.ustadmobile.core.controller.CommonHandlerPresenter
import com.ustadmobile.core.db.dao.PersonPictureDao
import com.ustadmobile.core.impl.UmAccountManager
import com.ustadmobile.lib.db.entities.Person
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import java.io.File

/**
 * A Simple recycler adapter for a dead simple list of Students.
 */

class SimplePeopleListRecyclerAdapter : PagedListAdapter<Person,
        SimplePeopleListRecyclerAdapter.ClazzStudentViewHolder> {

    internal var theContext: Context
    internal lateinit var theFragment: Fragment
    internal lateinit var mPresenter: CommonHandlerPresenter<*>
    private var personPictureDaoRepo: PersonPictureDao?=null

    class ClazzStudentViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    protected constructor(diffCallback: DiffUtil.ItemCallback<Person>,
                          context: Context, fragment: Fragment) : super(diffCallback) {
        theContext = context
        theFragment = fragment
    }

    constructor(diffCallback: DiffUtil.ItemCallback<Person>,
                          context: Context, fragment: Fragment,
                          presenter: CommonHandlerPresenter<*>) : super(diffCallback) {
        theContext = context
        theFragment = fragment
        mPresenter = presenter
    }

    protected constructor(diffCallback: DiffUtil.ItemCallback<Person>,
                          context: Context) : super(diffCallback) {
        theContext = context
    }

    constructor(diffCallback: DiffUtil.ItemCallback<Person>,
                          context: Context, presenter: CommonHandlerPresenter<*>)
            : super(diffCallback) {
        theContext = context
        mPresenter = presenter
    }

    /**
     * This method inflates the card layout (to parent view given) and returns it.
     * @param parent View given.
     * @param viewType View Type not used here.
     * @return New ViewHolder for the ClazzStudent type
     */
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ClazzStudentViewHolder {

        val clazzStudentListItem = LayoutInflater.from(theContext).inflate(
                R.layout.item_peoplelist, parent, false)
        return ClazzStudentViewHolder(clazzStudentListItem)
    }

    /**
     * This method sets the elements after it has been obtained for that item'th position.
     * @param holder    The holder
     * @param position  The position in the recycler view.
     */
    override fun onBindViewHolder(
            holder: ClazzStudentViewHolder, position: Int) {


        val personPicture = holder.itemView.findViewById<ImageView>(R.id.item_peoplelist_imageview)
        val thisPerson = getItem(position)
        val studentName: String
        if (thisPerson == null) {
            studentName = "Student"
        } else {
            studentName = thisPerson.firstNames + " " + thisPerson.lastName
        }

        val studentEntry = holder.itemView
                .findViewById(R.id.item_peoplelist_name) as TextView
        studentEntry.text = studentName

        studentEntry.setOnClickListener {
            v -> mPresenter.handleCommonPressed(thisPerson!!.personUid) }

        var imgPath = ""
        GlobalScope.async(Dispatchers.Main) {

            personPictureDaoRepo =
                    UmAccountManager.getRepositoryForActiveAccount(theContext).personPictureDao
            val personPictureDao = UmAccountManager.getActiveDatabase(theContext).personPictureDao

            val personPictureLocal = personPictureDao.findByPersonUidAsync(thisPerson!!.personUid)
            imgPath = personPictureDaoRepo!!.getAttachmentPath(personPictureLocal!!)!!

            if (!imgPath.isEmpty())
                setPictureOnView(imgPath, personPicture!!)
            else
                personPicture.setImageResource(R.drawable.ic_person_black_new_24dp)

            val personPictureEntity = personPictureDaoRepo!!.findByPersonUidAsync(
                    thisPerson.personUid)
            imgPath = personPictureDaoRepo!!.getAttachmentPath(personPictureEntity!!)!!

            if(personPictureLocal != personPictureEntity) {
                if (!imgPath.isEmpty())
                    setPictureOnView(imgPath, personPicture!!)
                else
                    personPicture.setImageResource(R.drawable.ic_person_black_new_24dp)
            }
        }
    }

    private fun setPictureOnView(imagePath: String, theImage: ImageView) {

        val imageUri = Uri.fromFile(File(imagePath))

        Picasso
                .get()
                .load(imageUri)
                .resize(dpToPxImagePerson(), dpToPxImagePerson())
                .noFade()
                .into(theImage)
    }

    companion object {

        private val IMAGE_PERSON_THUMBNAIL_WIDTH = 26

        private fun dpToPxImagePerson(): Int {
            return (IMAGE_PERSON_THUMBNAIL_WIDTH *
                    Resources.getSystem().displayMetrics.density).toInt()
        }
    }
}
